coding_guidelines.md

```diff
@@ -0,0 +1,500 @@
+# Table of Contents
+1. [Types](#types)
+2. [References](#references)
+3. [Objects](#objects)
+4. [Arrays](#arrays)
+5. [Functions](#functions)
+6. [Arrow Functions](#arrow-functions)
+7. [Destructuring](#destructuring)
+8. [Strings](#strings)
+
+# Types
+### [Primitives](#primitives)
+Primitives are passed by value.
+```typescript
+const favoriteNumber: number = 88;
+let newFavoriteNumber: number = favoriteNumber;
+newFavoriteNumber = 42;
+console.log(favoriteNumber, newFavoriteNumber); // => 88, 42
+```
+#### Primitive types
+- `string`
+- `number`
+- `boolean`
+- `null`
+- `undefined`
+
+### [Complex](#complex)
+Complex types are passed by reference.
+```typescript
+const favoriteNumberArray: number[] = [88, 42];
+const newFavoriteNumberArray: number[] = favoriteNumberArray;
+newFavoriteNumberArray[1] = 1337;
+console.log(favoriteNumberArray[1], newFavoriteNumberArrayp[1]); // => 1337, 1337
+```
+#### Complex types
+- `object`
+- `array`
+- `function`
+
+# References
+Mosst reference rules are defined in our linter, and will give lint errors if misshandled.
+> **Note:** Please make sure to include the type on all references, e.g.
+```typescript
+const months: number = 12;
+let currentMonth: number = 6;
+```
+### [Immutable References](#imutable-references)
+Use `const` instead of `var` for all of your immutable references.
+### [Mutable References](#mutable-references)
+Use `let` instead of `var` if you need a mutable reference.
+> `let` is block-scoped rather than function scoped
+
+# Objects
+### [Object Instansiation](#object-instansiation)
+
+#### Literal Syntax
+Literal syntax is prefered when dealing with simple cases. Perfectly valid usage of literal syntax:
+```typescript
+interface User {
+  username: string;
+  age: number;
+}
+
+const user: User = {
+  username: 'kanedoe',
+  age: 42
+}
+```
+**Avoid!!!** using literal syntax if function are defined in the object!
+```typescript
+interface Store {
+  name: string;
+  isOpen(hour24: number): boolean;
+}
+
+const isOpen = (hour24: number) => {
+  return hour24 > 6 && hour24 < 20;
+}
+
+const butcherShop = {
+  name: 'Meaty',
+  isOpen // AVOID!
+}
```

Comment:
Why is this bad?

Reply:
This goes against the rule we set up that we want classes. Literal syntax is what we would use if we went for a more functional programming approach.